src/SAMDTCTimer.cpp

```diff
@@ -0,0 +1,442 @@
+
+    // disable the timer
+    disable();
+
+    bool inited = false;
+
+    for (int index = 0; index < TC_INST_NUM; index++)
+        if (instances[index])
+            inited = true;
+
+    if (!inited)
+    {
+        DMESG("SET APP");
+        tc_set_app_handler(tc_irq_handler);
+    }
+
+    instances[tc_index] = this;
+
+    setBitMode(BitMode32);
+
+    // configure
+    switch (bm)
+    {
+        case BitMode8:
+            tc->COUNT8.CTRLBCLR.bit.DIR = 1; // count up
+#ifdef SAMD21
+            while (tc->COUNT8.STATUS.bit.SYNCBUSY);
+#endif
+            break;
+        case BitMode16:
+            tc->COUNT16.CTRLBCLR.bit.DIR = 1; // count up
+#ifdef SAMD21
+            while (tc->COUNT16.STATUS.bit.SYNCBUSY);
+#endif
+            break;
+        case BitMode24:
+            // make the compiler shut up
+            break;
+        case BitMode32:
+            tc->COUNT32.CTRLBCLR.bit.DIR = 1; // count up
+#ifdef SAMD21
+            while (tc->COUNT32.STATUS.bit.SYNCBUSY);
+#endif
+            break;
+    }
+
+    // 1000 khz == 1 mhz
+    setClockSpeed(1000);
+}
+
+int SAMDTCTimer::enable()
+{
+    NVIC_SetPriority((IRQn_Type)this->irqN, 2);
+    NVIC_ClearPendingIRQ((IRQn_Type)this->irqN);
+    NVIC_EnableIRQ((IRQn_Type)this->irqN);
+    tc_set_enable(tc, true);
+    return DEVICE_OK;
+}
+
+int SAMDTCTimer::disable()
+{
+    NVIC_DisableIRQ((IRQn_Type)this->irqN);
+    tc_set_enable(tc, false);
+    return DEVICE_OK;
+}
+
+int SAMDTCTimer::reset()
+{
+    NVIC_DisableIRQ((IRQn_Type)this->irqN);
+    switch (bm)
+    {
+        case BitMode8:
+            tc->COUNT8.COUNT.reg = 0;
+#ifdef SAMD21
+            while (tc->COUNT8.STATUS.bit.SYNCBUSY);
+#endif
+            break;
+        case BitMode16:
+            tc->COUNT16.COUNT.reg = 0;
+#ifdef SAMD21
+            while (tc->COUNT16.STATUS.bit.SYNCBUSY);
+#endif
+            break;
+        case BitMode24:
+            // make the compiler shut up
+            break;
+        case BitMode32:
+            tc->COUNT32.COUNT.reg = 0;
+#ifdef SAMD21
+            while (tc->COUNT32.STATUS.bit.SYNCBUSY);
+#endif
+            break;
+    }
+    NVIC_EnableIRQ((IRQn_Type)this->irqN);
+    return DEVICE_OK;
+}
+
+int SAMDTCTimer::setMode(TimerMode t)
+{
+    // only support timer mode.
+    return DEVICE_OK;
+}
+
+int SAMDTCTimer::setCompare(uint8_t channel, uint32_t value)
+{
+    DMESG("SET");
+    if (channel > getChannelCount())
+        return DEVICE_INVALID_PARAMETER;
+
+    switch (bm)
+    {
+        case BitMode8:
+            tc->COUNT8.CC[channel].reg = value;
+#ifdef SAMD21
+            while (tc->COUNT8.STATUS.bit.SYNCBUSY);
+#endif
+            // add channel to MC0, MC0 is 4, MC1 is 5
+            tc->COUNT8.INTENSET.reg = (1 << (TC_INTENSET_MC0_Pos + channel));
+            break;
+        case BitMode16:
+            tc->COUNT16.CC[channel].reg = value;
+#ifdef SAMD21
+            while (tc->COUNT16.STATUS.bit.SYNCBUSY);
+#endif
```

Comment:
Maybe move these three wait functions to three macros and define them to nothing on 51?

Reply:
Cheers, will do